src/k_smallest.rs

```diff
@@ -1,20 +1,211 @@
-use alloc::collections::BinaryHeap;
-use core::cmp::Ord;
+use alloc::vec::IntoIter;
+use core::cmp::{Ord, Ordering, Reverse};
+use core::mem::replace;
 
-pub(crate) fn k_smallest<T: Ord, I: Iterator<Item = T>>(mut iter: I, k: usize) -> BinaryHeap<T> {
-    if k == 0 { return BinaryHeap::new(); }
+/// Consumes a given iterator, leaving the minimum elements in the provided storage in **ascending** order.
+fn k_smallest_general<T, I: Iterator<Item = T>>(
+    iter: I,
+    k: usize,
+    order: impl Fn(&T, &T) -> Ordering,
+) -> IntoIter<T> {
+    if k == 0 {
+        return Vec::new().into_iter();
+    }
+    let heap = MaxHeap::from_iter(k, order, iter);
+    heap.unwrap_sorted().into_iter()
+}
+
+pub(crate) fn reverse_cmp<T, F>(cmp: F) -> impl Fn(&T, &T) -> Ordering
+where
+    F: Fn(&T, &T) -> Ordering,
+{
+    move |a, b| cmp(b, a)
+}
+
+pub(crate) fn reverse_key<T, K, F>(f: F) -> impl Fn(&T) -> Reverse<K>
+where
+    F: Fn(&T) -> K,
+{
+    move |a| Reverse(f(a))
+}
+
+pub(crate) fn k_smallest<T, I>(iter: I, k: usize) -> IntoIter<T>
+where
+    T: Ord,
+    I: Iterator<Item = T>,
+{
+    k_smallest_general(iter, k, T::cmp)
+}
+
+pub(crate) fn k_smallest_by<T, I, F>(iter: I, k: usize, cmp: F) -> IntoIter<T>
+where
+    I: Iterator<Item = T>,
+    F: Fn(&T, &T) -> Ordering,
+{
+    k_smallest_general(iter, k, cmp)
+}
+
+pub(crate) fn k_smallest_by_key<T, I, F, K>(iter: I, k: usize, key: F) -> IntoIter<T>
+where
+    I: Iterator<Item = T>,
+    F: Fn(&T) -> K,
+    K: Ord,
+{
+    let iter = iter.map(|v| Pair(key(&v), v));
+
+    let results: Vec<_> = k_smallest_general(iter, k, Ord::cmp)
+        .map(|Pair(_, t)| t)
+        .collect();
+    results.into_iter()
```

Comment:
I guess the intent is to avoid multiple evaluations of `key` for the same argument. Please correct me if I'm mistaken.

We should do along the lines of `std` here (`sort_by_key` vs `sort_by_cached_key`), and simply allow multiple evaluations per element instead of allocating an auxiliary vector for intermediate results. If it's a problem, users can still employ your workaround of `map(|v| Pair(key(&v), v))` or use `k_smallest_by_key(iter, k, make_cached_key(key))` (again transferring the caching out of the algorithm, so that the caching could also be used for e.g. `min_set_by_key`), or `itertools` could - as a last resort - offer `k_smallest_by_cached_key`.

Reply:
Yeah, this set up is so that `key` only has to be evaluated once. 

However, the auxillary vector is only necessary to be consistent with the other methods returning `VecIntoIter`. If the return type can be an arbitrary iterable of T, then it could equally be a `Map<>` of the original `Vec<Pair>`, and no second allocation would be needed

Reply:
Just to be clear: The `Vec<T>` is fine for me, but not the `Vec<Pair<K, T>>`. Instead, this should call `k_smallest_general(original_unmapped_iter, k, |a,b| key(a).cmp(key(b))` - and not force all users to allocate an intermediate `Vec<Pair<K, T>>`.